### PR TITLE
Add content type and length in header

### LINKED ISSUE #13740: docker 20.10 pull proxy cache image failed: Error response from daemon: missing or empty Content-Type header
If you are reporting a problem, please make sure the following information are provided:

**Expected behavior and actual behavior:**

should pull proxy cache repo images success.

but docker 20.10 pull proxy cache repo images got  error `missing or empty Content-Type header`

**Steps to reproduce the problem:**
 
```bash
# docker.io is a proxy cache project
docker pull ${HARBOR_HOST}/docker.io/library/alpine:3.12
Error response from daemon: missing or empty Content-Type header
```
logs of docker desktop for mac

image

works well with docker 19.03.x
normal project works well

**Versions:**
Please specify the versions of following systems.

- harbor version: v2.1.1-5f52168e
- docker engine version: 20.10

### PR DESCRIPTION
**Pull request by stonezdj**
Fixes #13740
Update ManifestExist to return Descriptor instead of digest

Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/controller/proxy/controller.go [hunk 1]
```diff
@@ -65,31 +65,54 @@ type Controller interface {
...
 	})
 
 	return ctl
 }
 
+func (c *controller) EnsureTag(ctx context.Context, art lib.ArtifactInfo, tagName string) error {
+	// search the digest in cache and query with trimmed digest
+	var trimmedDigest string
+	err := c.cache.Fetch(TrimmedManifestlist+art.Digest, &trimmedDigest)
+	if err != cache.ErrNotFound {
```
Comment (reasonerjt) [inline src/controller/proxy/controller.go:102]: why other errors are accepted?

File: src/controller/proxy/manifestcache.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	libCache "github.com/goharbor/harbor/src/lib/cache"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"strings"
+	"time"
+)
+
+const defaultHandler = "default_handler"
```
Comment (reasonerjt) [inline src/controller/proxy/manifestcache.go:33]: should be "default"?

File: src/controller/proxy/manifestcache.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+		schema2.MediaTypeManifest:          &ManifestCache{local},
+		"default":                          &ManifestCache{local},
+	}
+	return registry
+}
+
+// ManifestCacheHandler define how to cache manifest content
+type ManifestCacheHandler interface {
+	// CacheContent - cache the content of the manifest
+	CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, contType string, r RemoteInterface)
```
Comment (reasonerjt) [inline src/controller/proxy/manifestcache.go:49]: is the parm `contType` needed?

File: src/controller/proxy/manifestcache.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+// CacheContent ...
+func (m *ManifestListCache) CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, contType string, r RemoteInterface) {
+	_, payload, err := man.Payload()
+	if err != nil {
+		log.Errorf("failed to get payload, error %v", err)
+		return
+	}
+	key := getManifestListKey(art.Repository, art.Digest)
+	log.Debugf("ache manifest list with key=cache:%v", key)
```
Comment (reasonerjt) [inline src/controller/proxy/manifestcache.go:69]: typo

File: src/controller/proxy/manifestcache.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	err = m.cache.Save(key, payload, manifestListCacheInterval)
+	if err != nil {
+		log.Errorf("failed to cache payload, error %v", err)
+	}
+	err = m.Push(ctx, art.Repository, getReference(art), man)
+	if err != nil {
+		log.Errorf("error when push manifest list to local :%v", err)
+	}
+	return
```
Comment (reasonerjt) [inline src/controller/proxy/manifestcache.go:78]: You don't need this `return`

File: src/controller/proxy/manifestcache.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func (m *ManifestListCache) updateManifestList(ctx context.Context, repo string, manifest distribution.Manifest) (distribution.Manifest, error) {
```
Comment (reasonerjt) [inline src/controller/proxy/manifestcache.go:81]: This is dup with `ListProcessor.UpdateManifestList`

File: src/controller/proxy/manifestcache.go (hunk 3 above)
Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:49]: not required, remove it

File: src/controller/proxy/manifestcache.go (hunk 4 above)
Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:69]: updated

File: src/controller/proxy/manifestcache.go (hunk 5 above)
Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:78]: removed

File: src/controller/proxy/controller.go (hunk 1 above)
Comment (stonezdj) [reply src/controller/proxy/controller.go:102]: Updated, other error should returned

File: src/controller/proxy/manifestcache.go (hunk 6 above)
Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:81]: removed

**Review by ywk253100 — APPROVED** [review]

File: src/controller/proxy/controller.go [hunk 7]
```diff
@@ -65,31 +65,56 @@ type Controller interface {
...
```
Comment (ywk253100) [inline src/controller/proxy/controller.go:102]: Seems the error handling logic isn't correct?
What happens if the error is `cache.ErrNotFound`?

File: src/server/middleware/repoproxy/proxy.go [hunk 8]
```diff
@@ -227,14 +230,31 @@ func DisableBlobAndManifestUploadMiddleware() func(http.Handler) http.Handler {
...
 	if err != nil {
 		return err
 	}
-	if !exist {
+	if !exist || desc == nil {
 		return errors.NotFoundError(fmt.Errorf("The tag %v:%v is not found", art.Repository, art.Tag))
 	}
-	w.Header().Set(dockerContentDigest, dig)
-	w.Header().Set(etag, dig)
+	go func() {
```
Comment (ywk253100) [inline src/server/middleware/repoproxy/proxy.go:240]: How about to consider the proxy in the request scope?
When we get a `HEAD` request, we can just proxy the content just like what we do for the `GET`

File: src/controller/proxy/manifestcache.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+// ManifestCacheHandler define how to cache manifest content
+type ManifestCacheHandler interface {
+	// CacheContent - cache the content of the manifest
+	CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, r RemoteInterface)
+	// Push - push the manifest to local registry
+	Push(ctx context.Context, repo, reference string, man distribution.Manifest) error
+}
+
+// ManifestListCache handle Manifest list type
+type ManifestListCache struct {
```
Comment (ywk253100) [inline src/controller/proxy/manifestcache.go:55]: Why there two separated implementation for manifest list and index? They are almost the same thing except the media type

File: src/controller/proxy/manifestcache.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+		schema2.MediaTypeManifest:          &ManifestCache{local},
+		"default":                          &ManifestCache{local},
+	}
+	return registry
+}
+
+// ManifestCacheHandler define how to cache manifest content
+type ManifestCacheHandler interface {
+	// CacheContent - cache the content of the manifest
+	CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, r RemoteInterface)
```
Comment (ywk253100) [inline src/controller/proxy/manifestcache.go:49]: The word `cache` of `CacheContent` is `proxy` not the redis cache, right?
If it is, rename the function to `ProxyContent` may be easier to understand

Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:49]: the cache also includes the redis cache

File: src/controller/proxy/manifestcache.go (hunk 9 above)
Comment (stonezdj) [reply src/controller/proxy/manifestcache.go:53]: The implementation of CacheContent are different.

File: src/controller/proxy/controller.go [hunk 11]
```diff
@@ -198,7 +223,7 @@ func (c *controller) ProxyManifest(ctx context.Context, art lib.ArtifactInfo, re
 
 	return man, nil
 }
-func (c *controller) HeadManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, string, error) {
+func (c *controller) HeadManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, *distribution.Descriptor, error) {
```
Comment (morlay) [inline src/controller/proxy/controller.go:226]: consider handle 404 return false.

Comment (morlay) [reply src/controller/proxy/controller.go:226]: upstream registry logs

```
{"err.code":"manifest unknown","err.detail":"unknown tag=apline","err.message":"manifest unknown","go.version":"go1.15.6","http.request.id":"30f25b93-0963-4594-844b-6d907225b3f0","http.request.method":"HEAD","http.request.remoteaddr":"172.16.0.144","http.request.uri":"/v2/library/nginx/manifests/apline","http.request.useragent":"Proxy-Cache/docker.io Proxy-Cache/quay.io Proxy-Cache/k8s.gcr.io harbor-registry-client","http.response.contenttype":"application/json","http.response.duration":"3.304635986s","http.response.status":404,"http.response.written":96,"level":"error","msg":"response completed with error","time":"2021-01-05T07:38:17.288885579Z","vars.name":"library/nginx","vars.reference":"apline"}
```

harbor core error:
```
[ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"http status code: 500, body: "}]} /harbor/src/pkg/registry/client.go:546, github.com/goharbor/harbor/src/pkg/registry.(*client).do/harbor/src/pkg/registry/client.go:253, github.com/goharbor/harbor/src/pkg/registry.(*client).ManifestExist/harbor/src/controller/proxy/remote.go:90, github.com/goharbor/harbor/src/controller/proxy.(*remoteHelper).ManifestExist/harbor/src/controller/proxy/controller.go:147, github.com/goharbor/harbor/src/controller/proxy.(*controller).UseLocalManifest/harbor/src/server/middleware/repoproxy/proxy.go:119, github.com/goharbor/harbor/src/server/middleware/repoproxy.handleManifest/harbor/src/server/middleware/repoproxy/proxy.go:99, github.com/goharbor/harbor/src/server/middleware/repoproxy.ManifestMiddleware.func1/harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1/usr/local/go/src/net/http/server.go:2042, net/http.HandlerFunc.ServeHTTP/harbor/src/server/middleware/v2auth/auth.go:163, github.com/goharbor/harbor/src/server/middleware/v2auth.Middleware.func2.1
```

File: src/controller/proxy/controller.go (hunk 7 above)
Comment (stonezdj) [reply src/controller/proxy/controller.go]: updated

**Review by ywk253100 — APPROVED** [review]

**Review by reasonerjt — APPROVED** [review]

**Pull request merged**